common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResultFactory.java

```diff
@@ -0,0 +1,28 @@
+package com.microsoft.identity.common.internal.providers.oauth2;
+
+import android.content.Intent;
+
+/**
+ * Abstract Factory class to construct {@link AuthorizationResult}
+ */
+
+public abstract class AuthorizationResultFactory {
```

Comment:
We can squeeze better type safety out of this class using a generic template.

Consider the following declaration for this class

```java
public abstract class AuthorizationResultFactory<GenericAuthorizationResult extends AuthorizationResult> {
    // variables, javadocs and stuff...
    public abstract GenericAuthorizationResult createAuthorizationResult(final int resultCode, final Intent data);
}
```

This will allow you to remove the `instanceof` checks inside of your unit tests and hence, no more type casting